prov/util/src/rocr_mem_monitor.c

```diff
@@ -0,0 +1,378 @@
+			ofi_hsa_status_to_string(hsa_ret));
+
+		ret = -FI_EIO;
+		goto err_rbmap_delete_entry;
+	}
+
+	dlist_insert_tail(&(*entry)->entry, &rocr_mm.entry_list);
+
+	FI_DBG(&core_prov, FI_LOG_MR,
+	       "ROCR buffer address %p length %lu monitor entry allocated\n",
+	       hsa_info.agentBaseAddress, hsa_info.sizeInBytes);
+
+	return FI_SUCCESS;
+
+err_rbmap_delete_entry:
+	ofi_rbmap_delete(&rocr_mm.entries, (*entry)->node);
+err_free_entry:
+	free(*entry);
+err:
+	*entry = NULL;
+	return ret;
+}
+
+static void rocr_mm_init(struct ofi_mem_monitor *monitor)
+{
+	dlist_init(&rocr_mm.entry_list);
+	ofi_rbmap_init(&rocr_mm.entries, rocr_rbmap_compare);
+	ofi_monitor_init(&rocr_mm.mm);
+}
+
+static void rocr_mm_cleanup(struct ofi_mem_monitor *monitor)
+{
+	assert(dlist_empty(&rocr_mm.entry_list));
+	ofi_monitor_cleanup(&rocr_mm.mm);
+}
+
+static int rocr_mm_start(struct ofi_mem_monitor *monitor)
+{
+	/* no-op */
+	return FI_SUCCESS;
+}
+
+static void rocr_mm_stop(struct ofi_mem_monitor *monitor)
+{
+	struct rocr_mm_entry *entry;
+
+	while (!dlist_empty(&rocr_mm.entry_list)) {
+		dlist_pop_front(&rocr_mm.entry_list, struct rocr_mm_entry,
+				entry, entry);
+		rocr_mm_entry_free(entry);
+	}
+}
+
+static void rocr_mm_unsubscribe(struct ofi_mem_monitor *monitor,
+				const void *addr, size_t len,
+				union ofi_mr_hmem_info *hmem_info)
+{
+	struct rocr_mm_entry *entry;
+
+	entry = rocr_mm_entry_find(addr, len);
+	if (entry) {
+		rocr_mm_entry_free(entry);
+
+		FI_DBG(&core_prov, FI_LOG_MR,
+		       "ROCR buffer address %p length %lu unsubscribed\n",
+		       addr, len);
+	}
+}
+
+/* Subscribe is designed to monitor an entire ROCR memory region even if the
+ * user provided region is smaller. All ROCR memory regions are inserted into
+ * an RB tree for tracking. Future subscriptions will always reuse RB tree
+ * entries if possible.
+ *
+ * RB tree entries can be removed in two different ways:
+ * 1. An unsubscribe against the memory region occurs. This will occur when ROCR
+ *    invokes the deregistration callback.
+ * 2. The ROCR memory monitor is stopped.
+ *
+ * Note: The ROCR memory monitor does not impose a limit on the number of ROCR
+ * memory regions which can be monitored.
+ */
+static int rocr_mm_subscribe(struct ofi_mem_monitor *monitor, const void *addr,
+			     size_t len, union ofi_mr_hmem_info *hmem_info)
+{
+	struct rocr_mm_entry *entry;
+	int ret = FI_SUCCESS;
+
+	entry = rocr_mm_entry_find(addr, len);
+	if (entry)
+		FI_DBG(&core_prov, FI_LOG_MR,
+		       "Reusing monitored ROCR buffer address %p length %lu\n",
+		       entry->iov.iov_base, entry->iov.iov_len);
+	else
+		ret = rocr_mm_entry_alloc(addr, len, &entry);
+
+
+	if (ret == FI_SUCCESS)
+		FI_DBG(&core_prov, FI_LOG_MR,
+		       "ROCR buffer address %p length %lu subscribed\n", addr,
+		       len);
+	else
+		FI_DBG(&core_prov, FI_LOG_MR,
+		       "ROCR buffer address %p length %lu subscribe failed: %s\n",
+		       addr, len, fi_strerror(-ret));
```

Comment:
You can combine these FI_DBG statements into one that just prints the strerror return.  On success, it will just say success.

Reply:
Will do.